org.eclipse.xtext.xbase.testing/src/org/eclipse/xtext/xbase/testing/OnTheFlyJavaCompiler2.java

```diff
@@ -0,0 +1,202 @@
+import org.eclipse.xtext.util.JavaVersion;
+import org.eclipse.xtext.xbase.lib.Conversions;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.eclipse.xtext.xbase.lib.Functions.Function0;
+import org.eclipse.xtext.xbase.lib.IterableExtensions;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Iterables;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+/**
+ * @author Sven Efftinge - Initial contribution and API
+ * @noextend This class is not intended to be subclassed by clients.
+ * @noreference This class is not intended to be referenced by clients.
+ * 
+ * @since 2.9
+ */
+@Singleton
+public class OnTheFlyJavaCompiler2 {
+	private InMemoryJavaCompiler inMemoryCompiler;
+
+	private JavaVersion javaVersion;
+
+	/**
+	 * Creates a new OnTheFlyCompiler that accepts Java6 compliant code.
+	 */
+	@Inject
+	public OnTheFlyJavaCompiler2(ClassLoader scope) {
+		this(scope, JavaVersion.JAVA6);
+	}
+
+	public OnTheFlyJavaCompiler2(ClassLoader scope, JavaVersion version) {
+		this.inMemoryCompiler = new InMemoryJavaCompiler(scope, version);
+		this.javaVersion = version;
+	}
+
+	/**
+	 * @since 2.11
+	 */
+	public void setJavaVersion(JavaVersion version) {
+		inMemoryCompiler.setJavaVersion(version);
+		javaVersion = version;
+	}
+
+	/**
+	 * @since 2.11
+	 */
+	public JavaVersion getJavaVersion() {
+		return javaVersion;
+	}
+
+	public Class<?> compileToClass(String classname, String code) {
+		InMemoryJavaCompiler.Result result = inMemoryCompiler.compile(new JavaSource(toJavaFile(classname), code));
+		try {
+			if (Iterables.any(result.getCompilationProblems(), CategorizedProblem::isError)) {
+				StringConcatenation message = new StringConcatenation();
+				message.append("Java code compiled with errors:");
+				message.newLine();
+				message.append(Joiner.on("\n").join(
+						Iterables.filter(result.getCompilationProblems(), CategorizedProblem::isError)));
+				message.newLineIfNotEmpty();
+				message.newLine();
+				message.append("Code was:");
+				message.newLine();
+				message.append(code);
+				message.newLineIfNotEmpty();
+				throw new IllegalArgumentException(message.toString());
+			}
+			return result.getClassLoader().loadClass(classname);
+		} catch (Throwable t) {
+			if (t instanceof ClassNotFoundException) {
+				ClassNotFoundException e = (ClassNotFoundException) t;
+				StringConcatenation message = new StringConcatenation();
+				message.append("Couldn\'t load \'");
+				message.append(classname);
+				message.append("\' ");
+				message.newLineIfNotEmpty();
+				message.append("source :");
+				message.newLine();
+				message.append("\t");
+				message.append(code, "\t");
+				message.newLineIfNotEmpty();
+				message.newLine();
+				message.append("PROBLEMS : ");
+				message.newLine();
+				message.append("\t");
+				message.append(Joiner.on("\n").join(result.getCompilationProblems()), "\t");
+				message.newLineIfNotEmpty();
```

Comment:
extract to method, sequence is used twice

Reply:
:(